File: onboardUser/scripts/onboard_user.mjs
import fs from 'fs';
import { generateRSAKeyPair, decryptRSA, sign } from '../../tools/js/crypto.js';
import solc from 'solc';
import Web3 from 'web3';

const SOLC_VERSION = '0.8.19';
const FILE_PATH = 'GetUserKeyContract.sol';
const PROVIDER_URL = 'http://localhost:7000';

function checkInstalledSolcVersion(solcVersion) {
    const installedVersion = solc.version();
    console.log('Installed Solc version:', installedVersion);
    if (!installedVersion.includes(solcVersion)) {
        throw new Error(`Solc version ${solcVersion} is not installed. Please install it.`);
    } else {
        console.log(`Solc version ${solcVersion} is already installed.`);
    }
}

function readSolidityCode(filePath) {
    if (!fs.existsSync(filePath)) {
        throw new Error(`The file ${filePath} does not exist`);
    }
    return fs.readFileSync(filePath, 'utf8');
}

function compileSolidity(solcVersion, filePath, solidityCode, mpcInstPath, mpcCorePath) {
    const input = {
        language: 'Solidity',
        sources: {
            'MPCInst.sol': { content: fs.readFileSync(mpcInstPath, 'utf8') },
            'MpcCore.sol': { content: fs.readFileSync(mpcCorePath, 'utf8') },
            [filePath]: { content: solidityCode },
        },
        settings: {
            outputSelection: {
                '*': {
                    '*': ['abi', 'metadata', 'evm.bytecode', 'evm.bytecode.sourceMap'],
                },
            },
        },
    };
    const output = JSON.parse(solc.compile(JSON.stringify(input)));
    return output.contracts[filePath][filePath.split('.')[0]];
}

async function deployContract(web3, account, contractABI, contractBytecode, gas_limit=10000000, gas_price='30'){
    const contract = new web3.eth.Contract(contractABI);
    const deployTx = contract.deploy({
        data: contractBytecode
    });
    const deployOptions = {
        data: deployTx.encodeABI(),
        gas: gas_limit,
        from: account.address,
        gasPrice: web3.utils.toWei(gas_price.toString(), 'gwei') // Convert gas price to wei
    };

    try {
        const signedTx = await web3.eth.accounts.signTransaction(deployOptions, account.privateKey);
        const txHash = await new Promise((resolve, reject) => {
            web3.eth.sendSignedTransaction(signedTx.rawTransaction)
                .once('receipt', receipt => resolve(receipt.transactionHash))
                .once('error', reject);
        });
    
        const receipt = await web3.eth.getTransactionReceipt(txHash);
        console.log('Contract deployed at:', receipt.contractAddress);
    
        return new web3.eth.Contract(contractABI, receipt.contractAddress);
    } catch (error) {
        console.error('Error deploying contract:', error);
        throw error;
    }
    
}

async function executeTransaction(web3, account, contract, func, gas_limit=10000000, gas_price='30') {
    try {
        const nonce = await web3.eth.getTransactionCount(account.address);
        const transaction = await func.send({
            from: account.address,
            chainId: 50505050,
            gas: gas_limit,
            gasPrice: web3.utils.toWei(gas_price, 'gwei'),
            nonce: nonce,
        });

        const txnReceipt = await web3.eth.getTransactionReceipt(transaction.transactionHash);
        console.log(`Transaction successful with hash: ${txnReceipt.transactionHash}`);
        return txnReceipt;
    } catch (error) {
        console.error('Error executing transaction:', error);
        throw error;
    }
}

async function main() {
    checkInstalledSolcVersion(SOLC_VERSION);
    const solidityCode = readSolidityCode("../contracts/" + FILE_PATH);
    
    const mpcInstPath = '../../lib/solidity/MPCInst.sol';
    const mpcCorePath = '../../lib/solidity/MpcCore.sol';
    const compiledContract = compileSolidity(SOLC_VERSION, FILE_PATH, solidityCode, mpcInstPath, mpcCorePath);

    const web3 = new Web3(new Web3.providers.HttpProvider(PROVIDER_URL));
    if (!web3.eth.net.isListening()) {
        console.log("Failed to connect to the node.");
        return;
    }
    const SIGNING_KEY = process.env.SIGNING_KEY;
    web3.eth.accounts.wallet.add(SIGNING_KEY);
    const account = web3.eth.accounts.wallet[0];

    const contract = await deployContract(web3, account, compiledContract.abi, compiledContract.evm.bytecode.object);
    
    const { publicKey, privateKey } = generateRSAKeyPair();
    const signedEK = sign(publicKey, Buffer.from(SIGNING_KEY.slice(2), 'hex'));
    
    const func = contract.methods.getUserKey(publicKey, signedEK);
    await executeTransaction(web3, account, contract, func);
    
    const encryptedKey = await contract.methods.getSavedUserKey().call({ from: account.address });
    const buf = Buffer.from(encryptedKey.substring(2), 'hex');
    const decryptedAESKey = decryptRSA(privateKey, buf);
    console.log("user AES key:", decryptedAESKey.toString('hex'));

    fs.appendFileSync('../../.env', `export USER_KEY='${decryptedAESKey.toString('hex')}'\n`);
}

main();

File: onboardUser/scripts/python/onboard_user.py
import os
from soda_python_sdk import generate_rsa_keypair, sign, recover_user_key
from lib.python.soda_web3_helper import SodaWeb3Helper, parse_url_parameter
import logging

FILE_NAME = 'GetUserKeyContract.sol'
FILE_PATH = 'onboardUser/contracts/'

def remove_user_key_from_file(filename):
    # Read the content of the file
    with open(filename, 'r') as file:
        all_lines = file.readlines()

    # Filter out the lines containing "USER_KEY"
    temp_lines = list(filter(lambda line: "USER_KEY" not in line, all_lines))

    # Write the modified content back to the same file
    with open(filename, 'w') as file:
        for line in temp_lines:
            file.write(line)

def getUserKeyShares(account, contract, receipt):
    user_key_events = contract.events.UserKey().process_receipt(receipt)

    key_0_share = None
    key_1_share = None
    # Filter events for the specific address
    for event in user_key_events:
        if event['args']['_owner'].lower() == account.address.lower():
            key_0_share = event['args']['_keyShare0']
            key_1_share = event['args']['_keyShare1']
    
    if key_0_share is None or key_1_share is None:
        raise Exception("Failed to find the key shares of the account address in the transaction receipt.")

    return key_0_share, key_1_share

def main(provider_url: str):

    signing_key = os.environ.get('SIGNING_KEY')

    soda_helper = SodaWeb3Helper(signing_key, provider_url)

    # Compile the contract
    success = soda_helper.setup_contract(FILE_PATH + FILE_NAME, "onboard_user")
    if not success:
        raise Exception("Failed to set up the contract")

    # Deploy the contract
    receipt = soda_helper.deploy_contract("onboard_user", constructor_args=[])
    if receipt is None:
        raise Exception("Failed to deploy the contract")

    contract = soda_helper.get_contract("onboard_user")
    
    # Generate new RSA key pair
    private_key, public_key = generate_rsa_keypair()
    # Sign the public key
    signature = sign(public_key, bytes.fromhex(signing_key[2:]))

    # Call the getUserKey function to get the encrypted AES shares
    receipt = soda_helper.call_contract_transaction("onboard_user", "getUserKey", func_args=[public_key, signature])
    if receipt is None:
        raise Exception("Failed to call the transaction function")

    encryptedKey0, encryptedKey1 = getUserKeyShares(soda_helper.get_account(), contract, receipt)

    # Recover the aes key using the RSA private key and the given shares
    decrypted_aes_key = recover_user_key(private_key, encryptedKey0, encryptedKey1)
    
    # Remove the old USER_KEY from the .env file
    remove_user_key_from_file('.env')

    # Write the data to a .env file
    with open('.env', 'a') as f:
        f.write(f"export USER_KEY='{decrypted_aes_key.hex()}'\n")


if __name__ == "__main__":
    url = parse_url_parameter()
    if (url is not None):
        try:
            main(url)
        except Exception as e:
            logging.error("An error occurred: %s", e)
            raise e
    
